test/test_transformer.cc

```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
 }
+
+static isl::schedule_node tile(isl::schedule_node node,
+			       std::vector<int> tileSizes,
+  			       int defaultSize = 32) {
+  isl::space space = isl::manage(isl_schedule_node_band_get_space(node.get()));
+  unsigned dims = space.dim(isl::dim::set);
+  isl::multi_val sizes = isl::multi_val::zero(space);
+  for(unsigned i = 0; i < dims; ++i) {
+    int tileSize = i < tileSizes.size() ? tileSizes[i] : defaultSize;
+    sizes = sizes.set_val(i, isl::val(node.get_ctx(), tileSize));
+  }
+  node = 
+    isl::manage(isl_schedule_node_band_tile(node.release(), sizes.release()));
+  return node;
+}
+
+static isl::multi_union_pw_aff permuteDims(isl::multi_union_pw_aff ps, int firstDim,
+					   int secondDim) {
+  auto scheduleFirstDim = ps.get_union_pw_aff(firstDim);
+  auto scheduleSecondDim = ps.get_union_pw_aff(secondDim);
+  ps = ps.set_union_pw_aff(secondDim, scheduleFirstDim);
+  ps = ps.set_union_pw_aff(firstDim, scheduleSecondDim);
+  return ps;
+}
+
+TEST(Transformer, matmul) {
+
+  auto ctx = ScopedCtx(pet::allocCtx());
+  auto petScop = pet::Scop::parseFile(ctx, "inputs/1mmWithOutInitStmt.c");
+  auto scop = petScop.getScop();
+
+  auto dependences = computeAllDependences(scop);
+  scop.schedule =
+      mergeIfTilable(scop.schedule.get_root(), dependences).get_schedule();
+ 
+  isl::schedule_node root = scop.schedule.get_root();
+ 
+  using namespace matchers;
+  isl::schedule_node node;
+  // clang-format off
+  auto matcher = band(
+    [&node] (isl::schedule_node n) {
+      if(isl_schedule_node_band_n_member(n.get()) < 3)
+        return false;
+      else { node = n; return true; }
+    },
+    leaf());
+  // clang-format on
+
+  bool matched = false;
+  if(ScheduleNodeMatcher::isMatching(matcher, root.child(0))) {
+    matched = true;
+  } 
+
+  isl::union_map reads = scop.reads.curry();
+  isl::union_map writes = scop.mustWrites.curry();
+
+  auto _i = placeholder(ctx);
+  auto _j = placeholder(ctx);
+  auto _k = placeholder(ctx);
+  auto _ii = placeholder(ctx);
+  auto _jj = placeholder(ctx);
+
+  auto _A = arrayPlaceholder();
+  auto _B = arrayPlaceholder();
+  auto _C = arrayPlaceholder();
+
+  auto psRead = allOf(access(_A, _i, _j), 
+		      access(_B, _i, _k),
+		      access(_C, _k, _j));
+  auto matchedReads = 
+    match(reads, psRead);
+  //std::cout << "matched Reads " << matchedReads.size() << std::endl;
+  auto psWrite = allOf(access(_A, _ii, _jj));
+  auto matchedWrites =
+    match(writes, psWrite);
+  //std::cout << "matched Writes " << matchedWrites.size() << std::endl;
+
+  // tmp[j][i] = alpha * A[i][k] * B[k][j] + tmp[i][j]
+  // pass the checks. this is because we do not link
+  // read and write at the moment. Placeholder are _not_
+  // reused between different calls to allOf. We can overcome
+  // this inspecting the placeholder for the write and the read.
+  // They should be equal. 
+ 
+  matched = matched && (matchedWrites.size() == 1);
+  matched = matched && (matchedReads.size() == 1);
```

Comment:
The name `matchedReads` makes it slightly confusing.  It implies it contains the matched read relations.  I expect three of those, but you check there is only one.  (I know the variable contains _match descriptors_, indexed by placeholders.)  Maybe `readMatches` is a better name here.